pkg/qrcodegenerator/qrcode/enddevices/enddevices.go

```diff
@@ -0,0 +1,106 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enddevices
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator/qrcode"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var (
+	errCharacter = errors.DefineInvalidArgument("character", "invalid character `{r}`")
+	errNoJoinEUI = errors.DefineFailedPrecondition("no_join_eui", "no JoinEUI")
+	errNoDevEUI  = errors.DefineFailedPrecondition("no_dev_eui", "no DevEUI")
+	errFormat    = errors.DefineInvalidArgument("format", "invalid format")
+)
+
+// Format is a end device QR code format.
+type Format interface {
+	Format() *ttnpb.QRCodeFormat
+	New() Data
+}
+
+// Data represents end device QR code data.
+type Data interface {
+	qrcode.Data
+	Encode(*ttnpb.EndDevice) error
+	// FormatID returns the ID of the format used to parse the QR Code data.
+	FormatID() string
+	// EndDeviceTemplate returns the End Device Template corresponding to the QR code data.
+	EndDeviceTemplate() *ttnpb.EndDeviceTemplate
+}
+
+// Server provides methods for end device QR codes.
+type Server struct {
+	endDeviceFormats map[string]Format
+}
+
+// New returns a new Server.
+func New(ctx context.Context) *Server {
+	s := &Server{
+		endDeviceFormats: make(map[string]Format),
+	}
+
+	// Register known formats.
+	s.endDeviceFormats[formatIDLoRaAllianceTR005] = new(LoRaAllianceTR005Format)
+	s.endDeviceFormats[formatIDLoRaAllianceTR005Draft2] = new(LoRaAllianceTR005Draft2Format)
+	s.endDeviceFormats[formatIDLoRaAllianceTR005Draft3] = new(LoRaAllianceTR005Draft3Format)
```

Comment:
I still don't get why this isn't a package global initialized at init time

Reply:
You can even expose the package global with a nice godoc saying it's read-only, but write to it in tests init

Reply:
I see what you mean but I think the implementation and your suggestion are isomorphic.  
Unless there is a (functional/maintainability) issue with the implementation (which I don't really see), can we say that this is a reviewer nitpick and I prefer to keep my implementation?

Reply:
OK, but there's another issue with this, see below